kernel/common/enumerable.rb

```diff
@@ -424,21 +440,16 @@ def each_cons(num)
   end
 
   def each_slice(slice_size)
-    unless block_given?
-      enum = to_enum(:each_slice, slice_size)
+    n = Rubinius::Type.coerce_to_collection_index slice_size
+    raise ArgumentError, "invalid slice size: #{n}" if n <= 0
 
-      # For non lazy enumerators the enum size must be set to the size of the
-      # slice, _not_ the size of the containing enumerator.
-      unless enum.kind_of?(Enumerator::Lazy)
-        Rubinius.privately { enum.size = enum.count }
+    unless block_given?
+      size = Proc.new do
+        (enum_size = enumerator_size).nil? ? nil : (enum_size.to_f / n).ceil
```

Comment:
I'm not sure if it's 100% safe here to rely on a slice size always being of size `total_enum_size / slice_size`, granted I can't think of a possibility where this is not the case.